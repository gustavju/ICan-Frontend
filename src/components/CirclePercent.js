import React from 'react';

const CirclePercent = ({ percent, percentColor }) => {
    let percentNum = parseInt(percent);
    // 90 is from svg r (radius)
    let strokeDashoffset = ((100 - percentNum) / 100) * (Math.PI * (90 * 2));
    let stroke = percentColor;
    return (
        <div className="circle-percent__label" data-pct={percent}>
            <svg class="circle-percent__svg" width="200" height="200" viewport="0 0 100 100" version="1.1" xmlns="http://www.w3.org/2000/svg">
                <circle r="90" cx="100" cy="100" strokeDasharray="565.48" strokeDashoffset="0"></circle>
                <circle style={{strokeDashoffset, stroke}} r="90" cx="100" cy="100" strokeDasharray="565.48" strokeDashoffset="0"></circle>
            </svg>
        </div>
    );
}

export default CirclePercent;
