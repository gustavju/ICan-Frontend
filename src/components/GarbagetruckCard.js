import React from 'react';

const GarbagetruckCard = ({ selectGarbagetruck, garbagetruck }) => (
    <div onClick={(e) => selectGarbagetruck(e, garbagetruck.garbageTruckId)} className="garbagetruck-card">
        <div className="garbagetruck-card__id">{garbagetruck.garbageTruckId}</div>
        <div className="garbagetruck-card__text"><span>Capacity:</span>{garbagetruck.capacity}</div>
        <div className="garbagetruck-card__text"><span>Location:</span>{garbagetruck.location.longitude}, {garbagetruck.location.latitude}</div>
    </div>
);

export default GarbagetruckCard;