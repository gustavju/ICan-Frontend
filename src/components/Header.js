import React from 'react';
import { Link, NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { setTrashcans } from '../actions/trashcan';
import { setGarbagetrucks } from '../actions/garbagetruck';

class Header extends React.Component {
    constructor(props) {
        super(props);
        this.interval = setInterval(() => {
            this.updateTrash();
            this.updateGarbagetrucks();
        }, 3000);
    }
    componentWillUnmount() {
        clearInterval(this.interval);
    }
    updateTrash() {
        fetch('http://localhost:8500/getTrashcans').then(response => {
            response.json().then((data) => {
                console.log(data);
                this.props.setTrashcans(data);
            });
        });
    }
    updateGarbagetrucks() {
        fetch('http://localhost:8500/getGarbagetruck').then(response => {
            response.json().then((data) => {
                console.log(data);
                this.props.setGarbagetrucks(data);
            });
        });
    }
    render() {
        return (
            <header className="header">
                <div className="content-container">
                    <div className="header__content">
                        <Link className="header__title" to="/">
                            <h1>ICan</h1>
                        </Link>
                        <NavLink activeClassName="is-active" exact={true} className="header__link" to="/">
                            <h4>Trashcans</h4>
                        </NavLink>
                        <NavLink activeClassName="is-active" className="header__link" to="/garbagetruck">
                            <h4>Garbagetrucks</h4>
                        </NavLink>

                    </div>
                </div>
            </header>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        trashcans: state.trashcans,
        garbagetrucks: state.garbagetrucks,
    }
};
const mapDispatchToProps = (dispatch) => ({
    setTrashcans: (trashcans) => dispatch(setTrashcans(trashcans)),
    setGarbagetrucks: (garbagetrucks) => dispatch(setGarbagetrucks(garbagetrucks))
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
