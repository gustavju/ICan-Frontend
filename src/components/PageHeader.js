import React from 'react';

const PageHeader = ({ title, subTitle, data }) => (
    <div className="page-header">
        <div className="content-container">
            <h2 className="page-header__title">{title}</h2>
            <h3 className="page-header__subtitle"><span>{data}</span> {subTitle}</h3>
        </div>
    </div>
);

export default PageHeader;
