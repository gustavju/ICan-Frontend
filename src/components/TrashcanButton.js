import React from 'react';
import FontAwesome from 'react-fontawesome';
import { connect } from 'react-redux';
import { sendCommand } from '../actions/trashcan';

const commandToIcon = {
    empty: 'recycle',
    addTrash: 'plus',
    startTrashFire: 'fire'
};

export const TrashcanButton = ({ command, trashcanId, sendCommand }) => {
    return (
        <button className="trashcan-card__button"
            onClick={() => sendCommand(command, trashcanId)}>
            <FontAwesome name={commandToIcon[command]} />
        </button>
    );
}

const mapDispatchToProps = (dispatch) => ({
    sendCommand: (command, trashcanId) => dispatch(sendCommand(command, trashcanId))
});

export default connect(undefined, mapDispatchToProps)(TrashcanButton);