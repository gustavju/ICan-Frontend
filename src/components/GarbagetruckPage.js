import React from 'react';
import { connect } from 'react-redux';
import GoogleMapReact from 'google-map-react';
import FontAwesome from 'react-fontawesome';
import TrashcanTable from './TrashcanTable';
import PageHeader from './PageHeader';
import GarbagetruckCard from './GarbagetruckCard';
import { getPercentColor } from '../styles/styleFunctions';

class GarbagetruckPage extends React.Component {
    constructor() {
        super();
        this.state = {
            garbagetrucks: [],
            selectedTrashcans: [],
            selctedGarbagetruck: {},
            map: '',
            maps: ''
        };
        this.interval = setInterval(() => this.updateGarbagetrucks(), 3000);
        this.updateGarbagetrucks = this.updateGarbagetrucks.bind(this);
    }
    componentWillUnmount() {
        clearInterval(this.interval);
    }
    updateGarbagetrucks() {
        fetch('http://localhost:8500/getGarbagetruck').then(response => {
            response.json().then((data) => {
                console.log(data);
                this.setState(() => ({ garbagetrucks: data }));
            });
        });
    }
    sendRoute = () => {
        let trashcanQuery = '';
        const selectedTrashcans = this.state.selectedTrashcans;
        for (let i = 0; i < selectedTrashcans.length; i++) {
            trashcanQuery += `t${i}=${selectedTrashcans[i].trashcanId}&`;
        }
        const url = `http://localhost:8500/garbagetruckRoute?garbagetruckId=${this.state.selctedGarbagetruck.garbageTruckId}&num=${selectedTrashcans.length}&${trashcanQuery}`;
        console.log(url);
        fetch(url).then(res => {
            res.json().then((data) => console.log(data.Message));
        });
        this.getGoogleRoute();
    }
    selectTrashcan = (e, id) => {
        const element = e.target;
        if (!element.classList.contains('selected')) {
            element.classList.add("selected");
            let trashcan = this.props.trashcans.find(trashcan => trashcan.trashcanId === id);
            this.setState(() => ({ selectedTrashcans: [...this.state.selectedTrashcans, trashcan] }));
        } else {
            element.classList.remove("selected");
            let filterdTrashcans = this.state.selectedTrashcans.filter(trashcan => !trashcan.trashcanId === id);
            this.setState(() => ({ selectedTrashcans: filterdTrashcans }));
        }
    }
    selectGarbagetruck = (e, id) => {
        const element = e.target;
        if (!element.classList.contains('selected')) {
            element.classList.add("selected");
            let grabagetruck = this.state.garbagetrucks.find(garbagetruck => garbagetruck.garbageTruckId === id);
            this.setState(() => ({ selctedGarbagetruck: grabagetruck }));
        } else {
            element.classList.remove("selected");
            this.setState(() => ({ selctedGarbagetruck: '' }));
        }
    }
    getGoogleRoute = () => {
        let map = this.state.map;
        let maps = this.state.maps;
        const directionsService = new maps.DirectionsService();
        const directionsDisplay = new maps.DirectionsRenderer();
        let waypoints = this.state.selectedTrashcans.map(trashcan => ({ location: new google.maps.LatLng(trashcan.location.latitude, trashcan.location.longitude) }))
        directionsService.route({
            origin: new google.maps.LatLng(this.state.selctedGarbagetruck.location.latitude, this.state.selctedGarbagetruck.location.longitude),
            destination: new google.maps.LatLng(59.383689, 17.938740),
            waypoints: waypoints,
            optimizeWaypoints: true,
            travelMode: 'DRIVING'
        }, (response, status) => {
            if (status === 'OK') {
                directionsDisplay.setDirections(response);
                const routePolyline = new google.maps.Polyline({ path: response.routes[0].overview_path });
                routePolyline.setMap(map);
            } else {
                console.log('Directions request failed due to ' + status);
            }
        });
    };
    render() {
        return (
            <div>
                <PageHeader title="Garbagetrucks" subTitle={"Connected"} data={this.state.garbagetrucks.length} />
                <GoogleMapReact
                    style={{ width: '100%', height: '400px', position: 'relative' }}
                    bootstrapURLKeys={{ key: process.env.GOOGLE_MAPS_API_KEY }}
                    defaultCenter={{ lat: 59.407859, lng: 17.944644 }}
                    defaultZoom={13}
                    onGoogleApiLoaded={({ map, maps }) => this.setState({ map, maps })}
                    yesIWantToUseGoogleMapApiInternals
                >
                    {
                        this.state.garbagetrucks.length > 0 ?
                            this.state.garbagetrucks.map(garbagetruck => {
                                let markerStyles = {
                                    fontSize: '2rem',
                                    color: 'grey'
                                };
                                return (
                                    <FontAwesome style={markerStyles}
                                        lat={garbagetruck.location.latitude}
                                        lng={garbagetruck.location.longitude}
                                        text={garbagetruck.garbageTruckId}
                                        name="truck" >
                                    </FontAwesome>
                                );
                            }
                            ) : <p>No connected garbagetrucks!</p>
                    }
                    {
                        this.props.trashcans.length > 0 ?
                            this.props.trashcans.map(trashcan => {
                                let color = trashcan.isConnected == 'false' ? 'grey' : getPercentColor(trashcan.TrashcanHistoryEntry.trashLevel);
                                let markerStyles = {
                                    fontSize: '2rem',
                                    color
                                };
                                return (
                                    <FontAwesome style={markerStyles}
                                        key={trashcan.trashcanId}
                                        lat={trashcan.location.latitude}
                                        lng={trashcan.location.longitude}
                                        text={trashcan.trashcanId}
                                        name="map-marker" >
                                    </FontAwesome>
                                );
                            }
                            ) : <p>No connected trashcans!</p>
                    }
                </GoogleMapReact>
                <div className="content-container">
                    <button className="button" onClick={this.sendRoute}>Send</button>
                    <button className="button" onClick={this.updateGarbagetrucks}>Update</button>
                    <div>{this.state.garbagetrucks.map(garbagetruck => <GarbagetruckCard selectGarbagetruck={this.selectGarbagetruck} garbagetruck={garbagetruck} />)}</div>
                    <TrashcanTable selectTrashcan={this.selectTrashcan} trashcans={this.props.trashcans} />
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        trashcans: state.trashcans
    }
};

export default connect(mapStateToProps, undefined)(GarbagetruckPage);