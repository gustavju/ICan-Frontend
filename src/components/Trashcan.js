import React from 'react';
import TrashcanButton from './TrashcanButton';

const commands = [
    'empty',
    'addTrash',
    'startTrashFire'
];

const Trashcan = ({trashcan}) => (
    <div className="trashcan-card">
        <div className="trashcan-card__id">ID: {trashcan.trashcanId}</div>
        <div className="trashcan-card__trashlevel">Trashlevel: {trashcan.TrashcanHistoryEntry.trashLevel}%</div>
        <hr/>
        <div className="trashcan-card__temperature">Temperature: {trashcan.TrashcanHistoryEntry.temperature}Â°C</div>
        <div className="trashcan-card__text"><span>Location:</span> {trashcan.location.longitude}, {trashcan.location.latitude}</div>
        <div className="trashcan-card__text"><span>Last Emptied:</span> {trashcan.TrashcanHistoryEntry.lastEmptied}</div>
        <div className="trashcan-card__text"><span>Data updated:</span> {trashcan.TrashcanHistoryEntry.timestamp}</div>
        <div className="trashcan-card__button-container">
            {commands.map((command) => <TrashcanButton key={command} trashcanId={trashcan.trashcanId} command={command} />)}
        </div>
    </div>
);

export default Trashcan;
