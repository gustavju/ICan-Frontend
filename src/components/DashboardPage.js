import React from 'react';
import Trashcan from './Trashcan';
import PageHeader from './PageHeader';
import GoogleMapReact from 'google-map-react';
import FontAwesome from 'react-fontawesome';
import { connect } from 'react-redux';
import { getPercentColor } from '../styles/styleFunctions';

export class DashboardPage extends React.Component {
    constructor(props) {
        super(props);
    }
    handleMouseEnter() {

    }
    handleMouseLeave() {

    }
    render() {
        return (
            <div>
                <PageHeader title="Trashcans" subTitle={"Connected"} data={this.props.trashcans.length} />
                <GoogleMapReact
                    style={{ width: '100%', height: '400px', position: 'relative' }}
                    bootstrapURLKeys={{ key: process.env.GOOGLE_MAPS_API_KEY }}
                    defaultCenter={{ lat: 59.407859, lng: 17.944644 }}
                    defaultZoom={13}
                    onChildMouseEnter={this.handleMouseEnter}
                    onChildMouseLeave={this.handleMouseLeave}
                >
                    {
                        this.props.trashcans.length > 0 ?
                            this.props.trashcans.map(trashcan => {
                                let color = getPercentColor(trashcan.TrashcanHistoryEntry.trashLevel);
                                if (trashcan.isConnected == 'false') color = 'grey';
                                let markerStyles = {
                                    fontSize: '2rem',
                                    color
                                };
                                return (
                                    <FontAwesome style={markerStyles}
                                        key={trashcan.trashcanId}
                                        lat={trashcan.location.latitude}
                                        lng={trashcan.location.longitude}
                                        text={trashcan.trashcanId}
                                        name="map-marker" >
                                    </FontAwesome>
                                );
                            }
                            ) : <p>No connected trashcans!</p>
                    }
                </GoogleMapReact>
                <div className="content-container">
                    <div className="trashcan-card__container">
                        {
                            this.props.trashcans.length > 0 ?
                                this.props.trashcans.map(trashcan => {
                                    let color = getPercentColor(trashcan.TrashcanHistoryEntry.trashLevel);
                                    return (
                                        <Trashcan percentColor={color} key={trashcan.trashcanId} trashcan={trashcan} />
                                    )
                                })
                                :
                                <p>No connected trashcans!</p>
                        }
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        trashcans: state.trashcans
    }
};

export default connect(mapStateToProps, undefined)(DashboardPage);