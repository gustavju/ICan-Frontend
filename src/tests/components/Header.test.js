import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

let wrapper = <Header />;

test('should render Header correctly', () => {
    expect(wrapper).toMatchSnapshot();
});