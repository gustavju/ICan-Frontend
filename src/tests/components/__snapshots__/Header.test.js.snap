// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/"
      >
        <h1>
          ICan
        </h1>
      </Link>
      <NavLink
        activeClassName="is-active"
        aria-current="page"
        className="header__link"
        exact={true}
        to="/"
      >
        <h4>
          Trashcans
        </h4>
      </NavLink>
      <NavLink
        activeClassName="is-active"
        aria-current="page"
        className="header__link"
        to="/garbagetruck"
      >
        <h4>
          Garbagetrucks
        </h4>
      </NavLink>
    </div>
  </div>
</header>
`;
